edb/server/compiler_pool/server.py

```diff
@@ -0,0 +1,545 @@
+        if self._cache.pop(client_id, None):
+            self._invalidated_clients.append(client_id)
+        self._last_used_by_client.pop(client_id, None)
+
+    def invalidate_last(self, cache_size):
+        if len(self._cache) == cache_size:
+            client_id, _ = self._cache.popitem(last=True)
+            self._invalidated_clients.append(client_id)
+            self._last_used_by_client.pop(client_id, None)
+
+    def flush_invalidation(self):
+        rv, self._invalidated_clients = self._invalidated_clients, []
+        for client_id in rv:
+            self._cache.pop(client_id, None)
+        return rv
+
+    async def call(
+        self,
+        method_name,
+        *args,
+        sync_state=None,
+        msg=None,
+    ):
+        assert not self._closed
+
+        if self._con.is_closed():
+            raise RuntimeError(
+                "the connection to the compiler worker process is "
+                "unexpectedly closed"
+            )
+
+        if msg is None:
+            msg = pickle.dumps((method_name, args))
+        return await self._con.request(msg)
+
+
+class MultiSchemaPool(pool_mod.FixedPool):
+    _worker_class = Worker  # type: ignore
+    _worker_mod = "remote_worker"
+    _workers: typing.Dict[int, Worker]  # type: ignore
+    _clients: typing.Dict[int, ClientSchema]
+
+    def __init__(self, cache_size, **kwargs):
+        super().__init__(
+            dbindex=None,
+            backend_runtime_params=None,
+            std_schema=None,
+            refl_schema=None,
+            schema_class_layout=None,
+            **kwargs,
+        )
+        self._inited = asyncio.Event()
+        self._cache_size = cache_size
+        self._clients = {}
+
+    @functools.lru_cache(maxsize=None)
+    def _get_init_args(self):
+        init_args = (
+            self._backend_runtime_params,
+            self._std_schema,
+            self._refl_schema,
+            self._schema_class_layout,
+        )
+        pickled_args = pickle.dumps(init_args, -1)
+        return init_args, pickled_args
+
+    async def _attach_worker(self, pid: int):
+        if not self._running:
+            return
+        if not self._inited.is_set():
+            await self._inited.wait()
+        return await super()._attach_worker(pid)
+
+    async def _wait_ready(self):
+        pass
+
+    async def _init_server(self, client_id: int, init_args_pickled: bytes):
+        (
+            dbs,
+            backend_runtime_params,
+            std_schema,
+            refl_schema,
+            schema_class_layout,
+            global_schema,
+            system_config,
+        ) = pickle.loads(init_args_pickled)
+
+        if self._inited.is_set():
+            logger.debug("New client %d connected.", client_id)
+            if self._backend_runtime_params != backend_runtime_params:
+                raise state_mod.IncompatibleClient("backend_runtime_params")
+            if not self._std_schema.eq(std_schema):
```

Comment:
This check seems very expensive.  Why can't we use catalog_version or some other indicator of compatibility instead?

Reply:
It was a leftover issue from PoC. Using `catalog_version` makes sense, I'm pushing a fix.